wallet/src/Pos/Wallet/Web/Methods/Payment.hs

```diff
@@ -137,17 +145,26 @@ sendMoney
     -> NonEmpty (CId Addr, Coin)
     -> m CTx
 sendMoney SendActions{..} passphrase moneySource dstDistr = do
-    addrMetas' <- getMoneySourceAddresses moneySource
+    let srcWallet = getMoneySourceWallet moneySource
+    rootSk <- getSKById srcWallet
+    checkPassMatches passphrase rootSk `whenNothing`
+        throwM (RequestError "Passphrase doesn't match")
+
+    addrMetas'' <- getMoneySourceAddresses moneySource
+    utxo <- getCachedUtxo
+    let addrsWithMoney =
+            HS.fromList . map (addressToCId . txOutAddress . toaOut) $
+            toList utxo
+        addrMetas' = filter (flip HS.member addrsWithMoney . cwamId) addrMetas''
```

Comment:
So we grab the `getMoneySourceAddresses` as before, but now we filter the addresses by being in the set of non-zero utxo entries. This cuts out addresses we don't need to consider.